### DISCUSSION TITLE
Getting Value error while executing _generalized_psi function

### DISCUSSION THREAD

**Discussion by mukherjeearup** [Telecominfraproject/oopt-gnpy#459] [Q&A] [answered]
Hi,
Many thanks for answering my previous question, it helps. :)
Now I am trying to run Gnpy code (specifically Non linear parts like GN, GGN, SRS) in jypyter notebook. GN, SRS ran fine but I am getting one error while executing GGN.  To be specific, "_generalized_psi" function is failing in below line:

----> 8     f1_array = arange(pump_carrier.frequency - (pump_carrier.baud_rate * (1 + pump_carrier.roll_off) / 2),
                                               pump_carrier.frequency + (pump_carrier.baud_rate * (1 + pump_carrier.roll_off) / 2),
                                                     f_pump_resolution)

**ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()**

@AndreaDAmico Can you please help to fix this error?

**Comment by AndreaDAmico** [comment, marked as answer]
Hi,
That error is raised within the numpy method `arange` when at least one of the three inputs is an array instead of being a scalar. You should find out which of them is an array and why. This behavior has to depend on your specific usage (maybe a wrong format in the input files) as it is not the expected behavior (that function is covered within the GNPy tests).

**Reply by mukherjeearup** [reply]
Hi @AndreaDAmico ,
Thank you so much for your reply, yes I was able to find out which one of the input is array and corrected it. Now I am able to run GGN with SRS interchannel crosstalk considered in my Jupyter notebook. But, I am getting little different NLI values with this code compared to GnPy APP (https://gnpy.app/) in single hop 2 node linear network. Launch power and all the parameters are same in both the cases and I have loaded "sim params.json" file in the GnPy app with GGN and Raman flag on.
Is the GnPy app running latest GnPy version? Is there any code changes done with respect to GGN? Not able to understand the reason for this difference. GN model is giving exactly same result in both the cases.

For example, for 1 dBm launch power I am attaching Gnpy app results:
[Gnpy app.txt](https://github.com/Telecominfraproject/oopt-gnpy/files/10515656/Gnpy.app.txt)


From Jupyter notebook code I am getting below results for NLI for 76 channels in C band:

array([29.54136882, 29.0190271 , 28.7815241 , 28.63289322, 28.52723301,
       28.44676041, 28.37900599, 28.32361295, 28.27734606, 28.2380935 ,
       28.20440724, 28.17527123, 28.1499203 , 28.12777021, 28.10836627,
       28.09134792, 28.07642375, 28.06335018, 28.0519309 , 28.04200335,
       28.03343118, 28.02609869, 28.01990768, 28.01477399, 28.01062508,
       28.00739787, 28.00503625, 28.0034851 , 28.00270191, 28.00265104,
       28.00330205, 28.00462858, 28.00660858, 28.0092235 , 28.01245783,
       28.01629859, 28.02073409, 28.02574844, 28.0313351 , 28.03749125,
       28.04421553, 28.05150213, 28.05935558, 28.06778511, 28.07680362,
       28.08642756, 28.09667701, 28.10757567, 28.11915174, 28.13143799,
       28.14447127, 28.15828831, 28.17293967, 28.18848599, 28.20499841,
       28.22256016, 28.24126876, 28.2612387 , 28.2826058 , 28.30553183,
       28.3302102 , 28.35687053, 28.38580243, 28.41736847, 28.45202798,
       28.49037359, 28.53318839, 28.58154436, 28.63691279, 28.70142342,
       28.77833048, 28.86784077, 28.98248903, 29.1399429 , 29.38553108,
       29.90976958])

**Reply by mukherjeearup** [reply]
One more question in this regard, as per Gnpy code, GGN uses eq. 21 from arXiv: 1710.02225.
I am not able to exactly co-relate the referred equation with the implemented code in GGN method. Is there any further equation or simplification which GGN code uses? Can you please point me to the exact equtaion which GGN code uses?

**Reply by AndreaDAmico** [reply]
The GGN implementation in GNPy has been obtained starting from eq. 24 in arXiv: 1710.02225, where the frequency integrals are separated with respect the distinct  self- and cross-phase modulation contributions to the overall NLI (the SPM and XPM, respectively), while the FWM is neglected). Furthermore, the integration (by parts) in z is performed numerically in the method `gnpy.core.science_utils.NLISolver._generalized_rho_nli`.

**Reply by mukherjeearup** [reply]
great thanks @AndreaDAmico @ojnas. One more doubt. Is there any power balancing/tilt management done in the per channel fiber launch power while using GGN model? Because I am seeing that my Jupyter notebook results for NLI OSNR is matching with Gnpy app in the middle of C band channels and deviating more as I go towards both the edges. Possible reason could be: I am using same per channel signal power for all the channels while calculating NLI OSNR with GGN , i.e numerator is same for NLI OSNR equation for all the channels. Could you please confirm?

**Reply by ojnas** [reply]
No, there is no automatic tilt management in GNPy. It's difficult to say what causes the deviation. Maybe if you post your code, someone could figure it out.

**Reply by mukherjeearup** [reply]
Sure @ojnas , I am attaching my GGN ipynb file. Attaching the results from Gnpy app as well for same inputs.
Please note that I have verified SRS power profile (which is used as one of the inputs for GGN) with GnPy code and it is exactly matching.
[Desktop.zip](https://github.com/Telecominfraproject/oopt-gnpy/files/10555669/Desktop.zip)

Inputs to GGN model:
Start freq =  191.35
end freq = 196.1
spacing = 50 GHz
Per channel launch power into fiber = 0 dBm (same for all 96 channels)
Connector in = 0
connector out = 0
span type = SSMF
span length = 80 km
atten coeff = 0.2 dB/km
No raman pump, no lumped losses
n2 = 2.6e-20
effective_area = 83e-12
ref_wavelength = 1550e-9
c = 299792458
all channels baud rate  = 32 GBd
roll off  = 0.15
DEFAULT_SLOT_WIDTH_STEP = 12.5e9
dispersion_tolerance=1
phase_shift_tolerance=0.1
k_ref = 5
beta2_ref = 21.3e-27
delta_f_ref = 50e9
rs_ref = 32e9
dispersion = 1.67e-5

Can anyone please check the reason for NLI OSNR diff? @AndreaDAmico @ojnas

**Reply by mukherjeearup** [reply]
Hi @ojnas , @AndreaDAmico did you get a chance to look at my code?

**Reply by ojnas** [reply]
Hi @mukherjeearup, the main issue with your code is that you're setting beta2 = abs(beta2). Another thing is that you set cuts to all carriers, while I believe you run gnpy with default sim_params.json, which specifies a limited number of "computed_channels". A third thing is that you set slot_width to 37.5 GHz, while by default in GNPy it would be equal to spacing (but I think this probably has negligible impact).

**Reply by mukherjeearup** [reply]
Many thanks @ojnas . Now I ran my code for less number of cuts as mentioned in sim_params.json and NLI values are now closer :)

**Reply by ojnas** [reply]
Note that you can simply remove the "computed_channels" attribute from sim_params.json if you want cuts to be all carriers. If I do that, and make the other two changes in your code that I mentioned above, I get identical NLI with your code and with gnpy.app.

**Comment by ojnas** [comment]
Hi @mukherjeearup,
the results with gnpy.app should not be different. You can also verify by running the gnpy-transmission-example script included with gnpy, which should give the same results as gnpy.app, with the same inputs. One thing I noticed is that the results you attached were for 2 dBm launch power, while you wrote 1 dBm, but maybe that was just a typo.
